File: web/types/workflow.ts
/**
 * The shared workflow schema — the contract between the generator
 * (lib/generator.ts), the visualizer (components/WorkflowGraph.tsx) and the
 * CALL-E compiler (lib/compiler.ts). Treat this file as the source of truth:
 * any change here must be reflected in all three consumers.
 */

/** What a node does in the conversation. */
export type NodeType = "start" | "question" | "decision" | "terminal";

export interface WorkflowNode {
  id: string;
  type: NodeType;
  /** Short purpose label shown on the graph and in the node table. */
  label: string;
  /** The prompt text — what the agent says at this step. */
  say: string;
  /** Names of the fields this node is expected to capture. */
  captures: string[];
  /** Canvas position, in graph coordinates. */
  x: number;
  y: number;
}

export interface WorkflowEdge {
  id: string;
  from: string;
  to: string;
  /** Branch condition, e.g. "Yes", "Qualified". Null for an unconditional step. */
  condition: string | null;
}

export type QualificationOperator = "gte" | "lte" | "eq" | "in";

export interface QualificationRule {
  /** A field name that appears in some node's `captures`. */
  field: string;
  operator: QualificationOperator;
  value: string | number | string[];
  points: number;
}

/** Scoring logic that decides the outcome of a call. */
export interface Qualification {
  rules: QualificationRule[];
  /** Total score at or above which the call counts as qualified. */
  threshold: number;
}

/**
 * CALL-E's supported JSON Schema subset. Anything outside it is rejected server side
 * or silently nulls out the result, so the type cannot express it in the first place.
 * See TECHNICAL_ARCH.md section 4.5.
 */
export type OutcomeFieldType =
  | "string"
  | "number"
  | "integer"
  | "boolean"
  | "object"
  | "array";

/**
 * One field of the structured result CALL-E extracts at the end of a call.
 *
 * Deliberately cannot express `$ref`, `oneOf`, `anyOf`, `allOf`, recursive schemas, or
 * `additionalProperties: true` — CALL-E rejects all of them. Do not widen this type
 * without reading TECHNICAL_ARCH.md section 4.5 first.
 *
 * Reserved names: if these fields are ever compiled into a `recipient_result_schema`
 * (only relevant if a future feature batches identical-task calls), avoid CALL-E's
 * reserved recipient field names — `summary`, `status`, `transcript`, `call_id`, and any
 * timing-related name such as `started_at`, `completed_at` or `duration`.
 */
export interface OutcomeField {
  name: string;
  type: OutcomeFieldType;
  description?: string;
  /** Compiles to `enum`. */
  enumValues?: string[];
  /** Listed under the parent object's `required`. */
  required?: boolean;
  /** Simple `array.items` only — no tuple forms. Required when type is "array". */
  items?: OutcomeField;
  /** Nested object fields. Required when type is "object". */
  properties?: OutcomeField[];
}

/** The structured data shape returned once a call completes. */
export interface OutcomeSchema {
  fields: OutcomeField[];
  /** The permitted values of the call's next-step disposition. */
  nextStep: string[];
}

export interface Workflow {
  id: string;
  /** The overall purpose of the workflow. */
  goal: string;
  nodes: WorkflowNode[];
  edges: WorkflowEdge[];
  qualification: Qualification;
  outcomeSchema: OutcomeSchema;
}

File: engine/app/models/workflow.py
"""
The Python mirror of types/workflow.ts. That file is the source of truth for the
schema; if it changes, this file, engine/app/graph_validation.py and
engine/app/compiler.py need to change with it, same as the TypeScript consumers
listed in its own docstring.

Wire format is camelCase (matches the TS types and what the Next.js app sends/expects).
Field names are snake_case Python-side with camelCase aliases so both
`Workflow(**camel_case_dict)` and `Workflow(goal=...)` work.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

NodeType = Literal["start", "question", "decision", "terminal"]


class WorkflowNode(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    type: NodeType
    label: str
    say: str
    captures: list[str] = Field(default_factory=list)
    x: float
    y: float


class WorkflowEdge(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    from_: str = Field(alias="from")
    to: str
    condition: str | None = None


QualificationOperator = Literal["gte", "lte", "eq", "in"]


class QualificationRule(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    field: str
    operator: QualificationOperator
    value: str | float | list[str]
    points: float


class Qualification(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    rules: list[QualificationRule] = Field(default_factory=list)
    threshold: float


OutcomeFieldType = Literal["string", "number", "integer", "boolean", "object", "array"]


class OutcomeField(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    name: str
    type: OutcomeFieldType
    description: str | None = None
    enum_values: list[str] | None = Field(default=None, alias="enumValues")
    required: bool = False
    items: OutcomeField | None = None
    properties: list[OutcomeField] | None = None


OutcomeField.model_rebuild()


class OutcomeSchema(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    fields: list[OutcomeField] = Field(default_factory=list)
    next_step: list[str] = Field(default_factory=list, alias="nextStep")


class Workflow(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    goal: str
    nodes: list[WorkflowNode]
    edges: list[WorkflowEdge]
    qualification: Qualification
    outcome_schema: OutcomeSchema = Field(alias="outcomeSchema")
